i6_models/config.py

```diff
@@ -0,0 +1,56 @@
+"""
+Provides the base class for configurations, model configurations can be derived from this base class e.g:
+
+@dataclass
+class ExampleConfig(ModelConfiguration):
+    hidden_dim: int = 256
+    name: str = "Example Configuration"
+
+class ExampleModule(Module):
+    __init__(self, cfg: ExampleConfig)
+        self.hidden_dim = cfg.hidden_dim
+
+This config can then be used in the construction of the model to provide parameters.
+Approach is inspired by the Fairseq: https://github.com/facebookresearch/fairseq/blob/main/fairseq/dataclass/configs.py
+"""
+from typing import Any
+from dataclasses import dataclass, fields
+import typeguard
+
+
+@dataclass
+class ModelConfiguration:
+    """
+    Base dataclass that supports fetching attributes.
+    In order to enforce typing in a Config, decorate that Config with @enforce_types.
+    """
+
+    @classmethod
+    def from_namespace(cls, args: Any):
```

Comment:
Sorry to ask a bit late on this, but now I wonder: What is this function `from_namespace` actually used for?

Reply:
I would use it to initialize some configs with other existing ones. If I have a larger config with more parameters and want to initialize e.g. my encoder config with it, it can be done this way without having having to manually set all parameters again.

Reply:
Is this really a good idea? What is the benefit of doing this? I assume that you use this larger config for multiple different modules? But then I can imagine that there can be very nasty bugs or problems. E.g. when some option is maybe interpret differently depending on module (`dropout`, `momentum` and some others can be interpret as 1 - x or x), or some option name is not really consistent?

Didn't we actually discourage to do this for RASR, i.e. setting `*.momentum = 0.1` or so?

Can you add a docstring explaining this?

Reply:
I think also @michelwi wasn't a big fan of this either. This is what I would imagine to use it for, but if people rather prefer not to have this, due to the high chance of errors, then I can just remove it (and possible re-add it if we want this later on).
But then I would remove the Fairseq reference, because that was the last Fairseq part :D

Reply:
It's a bit difficult to discuss on this without seeing actual examples. Maybe you can find some in Fairseq?

@mmz33 do you know whether Zoltan uses this?

It still make sense to keep the Fairseq reference, but remove the "inspired from" part, just list it as a possible alternative, or extension. It might also make sense to list some other alternatives.

In general, it would be a good idea to have a short comparison to some of the popular alternatives. I know Lingvo also has such a configuration system (@mmz33 can maybe say sth about it). But I think there are also a few popular standalone packages for managing the configuration.

Btw, sorry if this question is a bit late: Did we actually look at some of these popular standalone packages, and did we consider to use just something existing, instead of doing our own thing here? Why did we decide against reusing sth existing?

Ok, when our `ModelConfiguration` is actually really minimal, like it is now, esp after removing `from_namespace`, it is not really needed.

Reply:
The Fairseq Transformer makes use of it but I feel its hart to follow the code because it not really documented, e.g:
https://github.com/facebookresearch/fairseq/blob/3f6ba43f07a6e9e2acf957fc24e57251a7a3f55c/fairseq/models/transformer/transformer_decoder.py#L467

In the first meeting we talked about this a bit and decided that we wanted to stay close to the style Fairseq does it.

Reply:
I rather meant examples of configurations where options are just put into a flat namespace. To see whether such configuration is easy to understand, or whether it rather makes things ambiguous.

Reply:
> Btw, sorry if this question is a bit late: Did we actually look at some of these popular standalone packages, and did we consider to use just something existing, instead of doing our own thing here? Why did we decide against reusing sth existing?

We did talk about it. We decided to base the config on the fairseq config system.

> I think also @michelwi wasn't a big fan of this either. 

Yes i am very much afraid that this leads to silent bugs where a user thinks they have sweet a config option but in fact it has been dropped or misinterpreted.

Reply:
I can see that. Also I could not really find an easy to read example, but then I never worked with Fairseq and so maybe I just dont know where to look. But I will remove it for now then.

Reply:
> We decided to base the config on the fairseq config system.

Ok, basing it on this, but not really using it directly. I really mean to use sth existing, not maintaining our own code for this.

But I also mean other independent packages which are just for this purpose.

What alternatives did you consider? It would be good to list alternatives also in the docstring, to understand later why some decision was made.